helper/cli.c

```diff
@@ -0,0 +1,484 @@
+					 const char *command ODP_UNUSED,
+					 char *argv[] ODP_UNUSED, int argc)
+{
+	if (check_num_args(cli, argc, 0))
+		return CLI_ERROR;
+
+	odp_sys_config_print();
+
+	return CLI_OK;
+}
+
+static int cmd_call_odp_sys_info_print(struct cli_def *cli,
+				       const char *command ODP_UNUSED,
+				       char *argv[] ODP_UNUSED, int argc)
+{
+	if (check_num_args(cli, argc, 0))
+		return CLI_ERROR;
+
+	odp_sys_info_print();
+
+	return CLI_OK;
+}
+
+static int cmd_call_odp_pktio_print(struct cli_def *cli,
+				    const char *command ODP_UNUSED,
+				    char *argv[], int argc)
+{
+	if (check_num_args(cli, argc, 1))
+		return CLI_ERROR;
+
+	odp_pktio_t hdl = odp_pktio_lookup(argv[0]);
+
+	if (hdl == ODP_PKTIO_INVALID) {
+		cli_error(cli, "%% Name not found.");
+		return CLI_ERROR;
+	}
+
+	odp_pktio_print(hdl);
+
+	return CLI_OK;
+}
+
+static int cmd_call_odp_pool_print(struct cli_def *cli,
+				   const char *command ODP_UNUSED, char *argv[],
+				   int argc)
+{
+	if (check_num_args(cli, argc, 1))
+		return CLI_ERROR;
+
+	odp_pool_t hdl = odp_pool_lookup(argv[0]);
+
+	if (hdl == ODP_POOL_INVALID) {
+		cli_error(cli, "%% Name not found.");
+		return CLI_ERROR;
+	}
+
+	odp_pool_print(hdl);
+
+	return CLI_OK;
+}
+
+static int cmd_call_odp_queue_print(struct cli_def *cli,
+				    const char *command ODP_UNUSED,
+				    char *argv[], int argc)
+{
+	if (check_num_args(cli, argc, 1))
+		return CLI_ERROR;
+
+	odp_queue_t hdl = odp_queue_lookup(argv[0]);
+
+	if (hdl == ODP_QUEUE_INVALID) {
+		cli_error(cli, "%% Name not found.");
+		return CLI_ERROR;
+	}
+
+	odp_queue_print(hdl);
+
+	return CLI_OK;
+}
+
+static int cmd_call_odp_shm_print(struct cli_def *cli,
+				  const char *command ODP_UNUSED, char *argv[],
+				  int argc)
+{
+	if (check_num_args(cli, argc, 1))
+		return CLI_ERROR;
+
+	odp_shm_t hdl = odp_shm_lookup(argv[0]);
+
+	if (hdl == ODP_SHM_INVALID) {
+		cli_error(cli, "%% Name not found.");
+		return CLI_ERROR;
+	}
+
+	odp_shm_print(hdl);
+
+	return CLI_OK;
+}
+
+static struct cli_def *create_cli(void)
+{
+	struct cli_command *c;
+	struct cli_def *cli;
+
+	cli = cli_init();
+	cli_set_banner(cli, NULL);
+	cli_set_hostname(cli, "ODP");
+
+	c = cli_register_command(cli, NULL, "show", NULL,
+				 PRIVILEGE_UNPRIVILEGED, MODE_EXEC,
+				 "Show information.");
+	cli_register_command(cli, c, "cpu", cmd_show_cpu,
+			     PRIVILEGE_UNPRIVILEGED, MODE_EXEC,
+			     "Show CPU information.");
+	cli_register_command(cli, c, "version", cmd_show_version,
+			     PRIVILEGE_UNPRIVILEGED, MODE_EXEC,
+			     "Show version information.");
+
+	c = cli_register_command(cli, NULL, "call", NULL,
+				 PRIVILEGE_UNPRIVILEGED, MODE_EXEC,
+				 "Call ODP API function.");
+	cli_register_command(cli, c, "odp_ipsec_print",
+			     cmd_call_odp_ipsec_print,
+			     PRIVILEGE_UNPRIVILEGED, MODE_EXEC, NULL);
+	cli_register_command(cli, c, "odp_pktio_print",
+			     cmd_call_odp_pktio_print,
+			     PRIVILEGE_UNPRIVILEGED, MODE_EXEC, NULL);
```

Comment:
Help strings are missing from the API calls. Also, in the help string it would be good to mention when an object name argument is required.

Reply:
The help string is not exactly the correct place for parameter help, but since there is no place for that in libcli 1.9 (while apparently there is a place in libcli 1.10), I will add the help string "<name>" to all the commands that have that parameter. It doesn't look right, but perhaps it's better than nothing.

With the help strings in place, due to a bug(?) in some versions of libcli, the user may see different error messages ("incomplete command" vs. "name not found") when typing "<command> ?", but I don't think this is a problem.